src/ring_pedersen.rs

```diff
@@ -0,0 +1,318 @@
+            modulus: modulus.clone(),
+            s,
+            t,
+        };
+        let proof = PiPrmProof::prove(
+            rng,
+            &PiPrmInput::new(scheme.clone()),
+            &PiPrmSecret::new(&lambda, phi_n),
+        )?;
+        Ok(Self { scheme, proof })
+    }
+
+    /// Verifies that the underlying [`RingPedersen`] commitment scheme was constructed correctly
+    /// according to the associated [`PiPrmProof`].
+    pub(crate) fn verify(&self) -> Result<()> {
+        self.proof.verify(&PiPrmInput::new(self.scheme.clone()))
+    }
+
+    /// Returns the underlying [`RingPedersen`] commitment scheme associated
+    /// with this object.
+    pub(crate) fn scheme(&self) -> &RingPedersen {
+        &self.scheme
+    }
+
+    /// Generates a [`VerifiedRingPedersen`] object from a random number generator for testing purposes.
+    #[cfg(test)]
+    pub(crate) fn gen(rng: &mut (impl RngCore + CryptoRng)) -> Result<Self> {
+        let (sk, _, _) = DecryptionKey::new(rng)?;
+        Self::extract(&sk, rng)
+    }
+}
+
+impl RingPedersen {
+    /// Returns the underlying modulus.
+    pub(crate) fn modulus(&self) -> &BigNumber {
+        &self.modulus
+    }
+
+    /// Returns the underlying `s` parameter.
+    pub(crate) fn s(&self) -> &BigNumber {
+        &self.s
+    }
+
+    /// Returns the underlying `t` parameter.
+    pub(crate) fn t(&self) -> &BigNumber {
+        &self.t
+    }
+
+    /// Constructs a [`RingPedersen`] object from its constituent parts for testing purposes.
+    #[cfg(test)]
+    pub(crate) fn from_parts(s: BigNumber, t: BigNumber, modulus: BigNumber) -> Self {
+        Self { modulus, s, t }
+    }
+
+    /// Produces commitment randomness sampled from `+/- 2^range * modulus * N`,
+    /// where `N` is the modulus of the commitment scheme.
+    pub(crate) fn commitment_randomness(
+        &self,
+        range: usize,
+        modulus: &BigNumber,
+        rng: &mut (impl RngCore + CryptoRng),
+    ) -> CommitmentRandomness {
+        let randomness = random_plusminus_scaled(rng, range, &(modulus * &self.modulus));
+        CommitmentRandomness(randomness)
+    }
+
+    /// Produces a commitment to `value`, which must be drawn
+    /// from `+/- 2^range`. The commitment randomness is sampled from
+    /// `+/- 2^range * N`, where `N` is the modulus of the commitment scheme.
+    pub(crate) fn commit(
+        &self,
+        value: &BigNumber,
+        range: usize,
+        rng: &mut (impl RngCore + CryptoRng),
+    ) -> (Commitment, CommitmentRandomness) {
+        let randomness = self.commitment_randomness(range, &BigNumber::one(), rng);
+        let com = self.reconstruct(value, randomness.as_masked());
+        (com, randomness)
+    }
+
+    /// Reconstructs a commitment to `value` using [`MaskedRandomness`] `randomness`.
+    /// The commitment is computed as in [`RingPedersen::commit`], except `randomness` is used
+    /// instead of it being generated by the commitment process.
```

Comment:
nits:
- `commit` doesn't explain what it's computing :) Please add the formula up there.
- newline between first and second sentence.